clang/lib/Frontend/DependencyFile.cpp

```diff
@@ -312,19 +316,26 @@ static void PrintFilename(raw_ostream &OS, StringRef Filename,
 }
 
 void DependencyFileGenerator::outputDependencyFile(DiagnosticsEngine &Diags) {
-  if (SeenMissingHeader) {
-    llvm::sys::fs::remove(OutputFile);
+  Expected<llvm::vfs::OutputFile> O = OutputBackend->createFile(
+      OutputFile,
+      llvm::vfs::OutputConfig().setTextWithCRLF().setNoAtomicWrite());
+
+  if (!O) {
+    Diags.Report(diag::err_fe_error_opening)
+        << OutputFile << toString(O.takeError());
     return;
   }
 
-  std::error_code EC;
-  llvm::raw_fd_ostream OS(OutputFile, EC, llvm::sys::fs::OF_TextWithCRLF);
-  if (EC) {
-    Diags.Report(diag::err_fe_error_opening) << OutputFile << EC.message();
+  if (SeenMissingHeader) {
+    consumeError(O->discard());
```

Comment:
I think, combined with the `setNoAtomicWrite()`, this effectively keeps side effects for an existing file the same:
- Previously, `remove()` would discard it if there was a missing header.
- Now, the file is always created, then `discard()` discards it if there was a missing header. (An extra system call (create + remove), but that seems fine.)

However, this seems quite subtle. I could see someone removing `setNoAtomicWrite` in the future and being confused that the old files weren't getting removed anymore.

Seems like it would be better to add an OutputConfig bit (`setDeleteExistingOnOpen()`) and OnDiskOutputBackend feature. If you don't have time for that, I suggest:
- Add a comment here explaining the subtle side effects you're reproducing.
- Include a FIXME for adding the OutputConfig feature.

Also, this is the first example of a place where it's *important* that we don't cache failures (since we don't have a way of caching "delete this output"). Can you expand the comment in `CompileJobCache` / cc1_main.cpp / wherever it is saying we're not caching failures that includes this example?

Alternatively, in theory this could be cached too. The CASOutputBackend could observe that the was opened+discarded with the DeleteExistingOnOpen feature and somehow save that info for replay. Even if you want to go that route, I still think there ought to be a comment that notes that this is not *currently* cacheable for replay.